Draft: Final python2 version

vahidm1993 opened the pull request:
Summary of pull request to merge in new features developed at Agile Robots AG.

* The list of changes can be checked in the changelog
* The internal pipeline for python2 and python3 runs
* Current version: 0.15.0
* Features and fixed include:
  * Memory profiling tooling
  * Fixes / implementations for:
    * https://github.com/DLR-RM/RAFCON/issues/851
    * https://github.com/DLR-RM/RAFCON/issues/841 
    * https://github.com/DLR-RM/RAFCON/issues/840
    * https://github.com/DLR-RM/RAFCON/issues/837
    * https://github.com/DLR-RM/RAFCON/issues/831

Diff to the old pull request:

* Making printings in 'LoggingView' asynchronous to avoid GUI freezing
* Defining a new GUI config called 'MAX_LOGGING_BUFFER_LINES' that determines the maximum lines of the logging buffer

Open TODOs before the pr can be merged:

- [ ] ...


 Open TODOs after the pr:

 - [ ] release missing versions on pypi

franzlst commented:
Ah, there is one question which I already had for the previews pull request: You mention "Support for custom designs", but I cannot find any code related to this. Is that really part of this PR?

vahidm1993 commented:
> Ah, there is one question which I already had for the previews pull request: You mention "Support for custom designs", but I cannot find any code related to this. Is that really part of this PR?

No we will add it in the next pull request. I have deleted it.

vahidm1993 commented:
@franzlst Point 1, 2, and 4 of the #858 comment got resolved in this branch.

However, I could not reproduce the 3rd point.

I moved '99_bottles_of_beer' somewhere else and tried to open '99_bottles_of_beer_in_library'. After the replacement of the missing library with a dummy state, I moved '99_bottles_of_beer' back to its place and tried to substitute the dummy state with '99_bottles_of_beer', and everything looks fine.

I guess the bug that you could somehow reproduce is related to 'substitute state' functionality and is not relevant to the better support of the missing libraries future.

franzlst commented:
Thank you for the fixes!

I now get a lot of `AssertionErrors`, which prevent logs from being shown in the LoggingView, caused by:
`logging_console.py, line 150, in split_text
assert isinstance(text_to_split, str)`
This only happens for Python 2, for Python 3 it works fine.

> However, I could not reproduce the 3rd point.

Steps to reproduce:
* create new state machine, add input data port with type float
* add generic/wait state, connect income, success outcome and input data port to root state
* save state machine and close
* rename wait folder on filesystem to wait_renamed
* open state machine again, abort search for library => dummy state will be used
* right click on dummy state => Substitute state => Choose generic/wait_renamed => click "As library"
* The data flow will get lost, the transitions sometimes remain, sometimes not (haven't found out in which cases)

> I guess the bug that you could somehow reproduce is related to 'substitute state' functionality and is not relevant to the better support of the missing libraries' feature.

Not sure. If I try the same with a library (not a dummy state), substitution works:
* create new state machine, add input data port with type float
* add generic/wait state, connect income, success outcome and input data port to root state
* save state machine and close
* _copy_ wait folder on filesystem, name it wait_copy
* open state machine again
* right click wait state => Substitute state with library => Choose generic/wait_copy => click "As library"
* All connections remain

franzlst commented on source/rafcon/core/storage/storage.py:
```diff
@@ -369,6 +373,48 @@ def load_state_machine_from_path(base_path, state_machine_id=None):
+                    same_level_states.append(same_level_state)
+            for same_level_state in same_level_states:
+                for data_flow in state.parent.data_flows.values():
+                    data_type = int
```
This `reconnect_data_flow` function does not handle scoped variables. This is probably also why `data_type` is here set to `int` by default, because in the following, it is only checked if the data flow is connected to an input or output port, not to a scoped variable. If also the last case was handled, it wouldn't be necessary to initialize `data_type` and `default_value` with values.

I also verified this again with the `wait` library. When connecting the input of this library to a scoped variable of the parental state and I rename the wait state, the data type of the input port of the dummy state is int instead of float.

vahidm1993 replied:
Nice catch! Fixed. Still, I initialize data_type & default_value just to be on the safe side.

franzlst replied:
Thank you for quickly fixing this!

It is a good style in software development, to always have an `else` case. There you can catch errors, and e.g. print an error message, as this is a case that is not expected, so might point to an error at another location in the code.

franzlst reviewed (changes requested):
Thank you again for your fixes. In the code I added a hopefully last comment which you should consider fixing :-)

franzlst reviewed (approved):
Awesome, so the pull request is ready from my side. Shell I merge it? Actually the honor is in your side @sebastian-brunner :-)
After merging, we also need a PyPI release, right?

vahidm1993 commented:
@franzlst Our pipeline for python 2 was failing, so I went through the errors and fixed the issues. We had to back to 'imp' instead of 'importlib' which is python 2 friendly. Now, our pipeline is passing for python 2 & 3.

franzlst commented:
Perfect!

sebastian-brunner reviewed (approved):
This looks good from my side. We discussed the changes internally already.

sebastian-brunner commented:
Ok thx for all the work @vahidm1993  and for your precious comments @franzlst .

Now the real show starts with releasing the first major. xD

sebastian-brunner commented:
Could you create a release on pypi @franzlst ?

franzlst commented:
> Could you create a release on pypi @franzlst ?

Done so! I created the release on PyPI, as well as [here on GitHub]8https://github.com/DLR-RM/RAFCON/releases/tag/0.15.4). The master and develop branches should now both point to the latest commit. Also the rebuilt of the GitHub pages has been triggered.

Thank you again for your contributions. I hope that I will find time next week for the next round of pull requests :-D
